File: src/main/java/nl/mpcjanssen/simpletask/remote/FileStoreInterface.java
package nl.mpcjanssen.simpletask.remote;

import android.app.Activity;
import android.support.annotation.Nullable;
import java.io.IOException;
import java.util.List;

/**
 * Interface definition of the storage backend used.
 *
 */
public interface FileStoreInterface {
    boolean isAuthenticated();
    List<String> loadTasksFromFile(String path, @Nullable BackupInterface backup, String eol)  throws IOException;
    void startLogin(Activity caller, int i);
    void logout();
    void browseForNewFile(Activity act, String path, FileSelectedListener listener, boolean txtOnly);
    void saveTasksToFile(String path, List<String> lines, @Nullable BackupInterface backup, String eol) throws IOException;
    void appendTaskToFile(String path, List<String> lines, String eol) throws IOException;

    int getType();
    void sync();
    String readFile(String file, FileReadListener fileRead) throws IOException;
    boolean supportsSync();
    boolean changesPending();

    boolean isLoading();

    interface FileSelectedListener {
        void fileSelected(String file);
    }
    interface FileChangeListener {
        void fileChanged(String newName);
    }

    interface FileReadListener {
        void fileRead(String contents);
    }
}

File: app/src/main/java/nl/mpcjanssen/simpletask/remote/FileStoreInterface.kt
package nl.mpcjanssen.simpletask.remote

import android.app.Activity
import nl.mpcjanssen.simpletask.TodoApplication
import nl.mpcjanssen.simpletask.util.broadcastFileSync
import java.io.File
import java.io.IOException

/**
 * Interface definition of the storage backend used.
 * All functions are run in a background thread and are properly queued to prevent
 * concurrent Todo list modification.
 * The implementers of this interface should throw exceptions if anything goes wrong. The logic for handling the
 * errors should be implemented in the users of the FileStore

 */
interface FileStoreInterface {
    val isAuthenticated: Boolean

    @Throws(IOException::class)
    fun loadTasksFromFile(path: String, eol: String): RemoteContents
    fun saveTasksToFile(path: String, lines: List<String>, eol: String) : String

    fun startLogin(caller: Activity)
    fun logout()

    @Throws(IOException::class)
    fun appendTaskToFile(path: String, lines: List<String>, eol: String)

    @Throws(IOException::class)
    fun readFile(file: String, fileRead: FileReadListener?): String
    fun writeFile(file: File, contents: String)

    fun supportsSync(): Boolean

    fun getWritePermission(act : Activity, activityResult: Int): Boolean


    val isOnline: Boolean

    interface FileSelectedListener {
        fun fileSelected(file: String)
    }

    interface FileChangeListener {
        fun fileChanged(newName: String?)
    }

    interface FileReadListener {
        fun fileRead(contents: String?)
    }

    // Retrieve the remote file version
    fun getRemoteVersion(filename: String): String?

    fun getDefaultPath(): String

    fun loadFileList(path: String, txtOnly: Boolean): List<FileEntry>

    // Allow the FileStore to signal that the remote
    // todoFile changed. Call this in the filestore code
    // to force file sync
    fun remoteTodoFileChanged() {
        broadcastFileSync(TodoApplication.app.localBroadCastManager)

    }

    companion object {
        val ROOT_DIR = "/"
        val PARENT_DIR = ".."

    }
}

data class RemoteContents(val remoteId: String, val contents: List<String>)
data class FileEntry(val name: String, val isFolder: Boolean)



